### ISSUE TITLE
Secure storage implementation with RPMB

### ISSUE THREAD

**Issue by SarahL21** [OP-TEE/optee_os#2457]

**Comment by jforissier** [comment]
If you want to access RPMB from a TA, all you need is make sure OP-TEE is compiled with CFG_RPMB_FS=y (it is by default), then use TEE_STORAGE_PRIVATE_RPMB instead of TEE_STORAGE_PRIVATE when calling the TEE Internal API functions (TEE_CreatePersistentObject(), etc.). See https://github.com/linaro-swg/optee_examples/blob/master/secure_storage/ta/secure_storage_ta.c for an example of how to use these functions (and please see the GlobalPlatform spec, too).

The RPMB storage is emulated by tee-supplicant. If you need to access real hardware (eMMC RPMB partition), then you need to unset RPMB_EMU when building optee_client.

**Comment by SarahL21** [comment]
Just to have a clear idea: while building Op-tee as in [https://github.com/OP-TEE/build](https://github.com/OP-TEE/build) , I would have both the trusted and the normal world on my SD card right? Or should I built the op_client independently?

**Comment by SarahL21** [comment]
When building optee as in here [https://github.com/OP-TEE/build](https://github.com/OP-TEE/build) , I get only the secure world right? Should I build the optee_client for the normal world on the same sd card? It's still quite blurry for me..

**Comment by jforissier** [comment]
Yes, build.git builds everything (secure and non-secure software).

**Comment by SarahL21** [comment]
Great ! And when booting it on the Pi, by default, we're on the normal world right? I know the switch between the two worlds is done through the secure monitor, but how can I access the secure world. For example, dirf.db is encrypted, its content is unreadable.. would it be readable in the secure world? Because I need to implement the secure storage with RPMB, and I don't know where I need to put and compile my files exactly. Logically, in the secure world?

**Comment by SarahL21** [comment]
Hello, 
While building the client, I have unset RPMB_EMU but there is also : 
```
ifeq ($(RPMB_EMU),1)
TEES_CFLAGS	+= -DRPMB_EMU=1
```
What does DRPMB_EMU refer to ? Do I have to modify this as well please?
Many thanks,

**Comment by jforissier** [comment]
`-DRPMB_EMU=1` is a command-line switch for the C compiler that will define the `RPMB_EMU` macro to `1` in the C source code. This will make `tee-supplicant` emulate a virtual RPMB partition in memory instead of accessing to the real hardware eMMC/SD card.

The `ifeq ($(RPMB_EMU),1)` will add this switch when the make variable `RPMB_EMU` is `1` (which it is by default, see [here](https://github.com/OP-TEE/optee_client/blob/3.2.0/tee-supplicant/Makefile#L7)).

So, to remove the switch (and consequently allow access to the real device), you should build with `make RPMB_EMU=` or comment out `RPMB_EMU := ` in [`tee-supplicant/Makefile`](https://github.com/OP-TEE/optee_client/blob/3.2.0/tee-supplicant/Makefile#L7).

> While building the client, I have unset RPMB_EMU but there is also :

Note that RPMB_EMU has no effect when building the client **application**, only when building `tee-supplicant` which is part of `optee_client`.

HTH.

**Comment by SarahL21** [comment]
I got it! Perfect explanation, many thanks! 
A last question before I close the issue please: I read somewhere (I really do not remember the link), that secure storage TAs are called/executed differently than other TAs (for hello world example, we just type hello_world on the normal world console and see the hello and goodbye logs on the secure console), but that is not the same with the secure storage TA: `secure_storage` and `optee_example_secure_storage` do not seem to work. 
Would you know where is that mentioned or point me to how we can run it please? 
Thank you and sorry for the bother.

**Comment by jforissier** [comment]
Looks like the `CMakeList.txt` is missing in https://github.com/linaro-swg/optee_examples/tree/master/secure_storage. You can try to write one based on the other applications.

**Comment by AparnaRGadhiya** [comment]
Hello,
I am working on RPMB Secure Storage Implementation. I created persistent object which are successfully written in secure storage. But I cannot see any /dev/mmc0blk partition in my Linux file System. How to be sure if the implementation is working fine. 
I will be really great if someone could help me understand this.
Thankyou.

**Comment by etienne-lms** [comment]
Optee_client deault emukates an RPMB define, with limitations.  You should disable the emulation.
See config switch `RPMB_EMU` in optee_client source tree.

Note that if there is no /dev/mmcblkXrpmb device file in Linux, tee-supplicant will fail to access the real eMMC and your test will likely fail. You need to resolve this. Check you Linux config and DT to be sure emmc device and rpmb support is enabled.
